core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java

```diff
@@ -0,0 +1,449 @@
+import java.util.Collections;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * A collection of tombstones for explicitly marking indices as deleted in the cluster state.
+ *
+ * The cluster state contains a list of index tombstones for indices that have been
+ * deleted in the cluster.  Because cluster states are processed asynchronously by
+ * nodes and a node could be removed from the cluster for a period of time, the
+ * tombstones remain in the cluster state for a fixed period of time, after which
+ * they are purged.
+ */
+final public class IndexGraveyard implements ToXContent, Diffable<IndexGraveyard> {
+
+    // X-Content key for index tombstones
+    public static final String TOMBSTONES_KEY = "index_tombstones";
+
+    // Setting for the maximum tombstones allowed in the cluster state;
+    // prevents the cluster state size from exploding too large, but it opens the
+    // very unlikely risk that if there are greater than MAX_TOMBSTONES index
+    // deletions while a node was offline, when it comes back online, it will have
+    // missed index deletions that it may need to process.
+    private static final int MAX_TOMBSTONES = 10000;
+
+    // Setting for the expiration window of tombstone entries.  Any tombstone entries
+    // that have been in the list longer than EXPIRATION_WINDOW will be purged on the
+    // next cluster update.
+    static final long EXPIRATION_WINDOW = 1000 * 60 * 60 * 24 * 14; // 14 days
+
+    // holds the index tombstones
+    private final Deque<Tombstone> tombstones;
+
+    public IndexGraveyard() {
+        tombstones = new ArrayDeque<>();
+    }
+
+    // copy constructor
+    public IndexGraveyard(final IndexGraveyard that) {
+        this.tombstones = new ArrayDeque<>(that.tombstones);
+    }
+
+    public IndexGraveyard(final StreamInput in) throws IOException {
+        final int queueSize = in.readVInt();
+        tombstones = new ArrayDeque<>(queueSize);
+        for (int i = 0; i < queueSize; i++) {
+            tombstones.add(new Tombstone(in));
+        }
+    }
+
+    // only used by fromXContent
+    private IndexGraveyard(final List<Tombstone> list) {
+        this.tombstones = new ArrayDeque<>(list);
+    }
+
+    // only used by IndexGraveyardDiff
+    private IndexGraveyard(final Deque<Tombstone> queue) {
+        this.tombstones = new ArrayDeque<>(queue);
+    }
+
+    /**
+     * Do they have the same index tombstones (deletions)?
+     */
+    public boolean isEquals(IndexGraveyard that) {
+        // array deque doesn't implement equals
+        return Objects.equals(new ArrayList<>(tombstones), new ArrayList<>(that.tombstones));
+    }
+
+    /**
+     * Add a deleted index to the list of tombstones in the cluster state.  Returns the
+     * purged tombstone if the size of the list grew too big, else return none.
+     */
+    public Optional<Tombstone> addTombstone(final Index index) {
```

Comment:
can we move all purging to the purge method, which will always be called by the builder? The down side will be not having logging, but if people want that they can put the cluster service in trace logging and see the changes.